extensions/vault/runtime/src/main/java/io/quarkus/vault/runtime/client/OkHttpVaultClient.java

```diff
@@ -146,33 +169,45 @@ public VaultTransitEncrypt rewrap(String token, String keyName, VaultTransitRewr
 
     // ---
 
+    protected <T> T delete(String path, String token, Object body, Class<T> resultClass, int expectedCode) {
+        Request request = builder(path, token).delete(requestBody(body)).build();
+        return exec(request, resultClass, expectedCode, true);
+    }
+
     protected <T> T post(String path, String token, Object body, Class<T> resultClass, int expectedCode) {
         Request request = builder(path, token).post(requestBody(body)).build();
-        return exec(request, resultClass, expectedCode);
+        return exec(request, resultClass, expectedCode, true);
     }
 
     protected <T> T post(String path, String token, Object body, Class<T> resultClass) {
         Request request = builder(path, token).post(requestBody(body)).build();
-        return exec(request, resultClass);
+        return exec(request, resultClass, true);
     }
 
     protected <T> T put(String path, String token, Object body, Class<T> resultClass) {
         Request request = builder(path, token).put(requestBody(body)).build();
-        return exec(request, resultClass);
+        return exec(request, resultClass, true);
     }
 
     protected <T> T get(String path, String token, Class<T> resultClass) {
+        return get(path, token, resultClass, true);
+    }
+
+    protected <T> T get(String path, String token, Class<T> resultClass, boolean throwOn404) {
         Request request = builder(path, token).get().build();
-        return exec(request, resultClass);
+        return exec(request, resultClass, throwOn404);
     }
 
-    private <T> T exec(Request request, Class<T> resultClass) {
-        return exec(request, resultClass, 200);
+    private <T> T exec(Request request, Class<T> resultClass, boolean throwOn404) {
+        return exec(request, resultClass, 200, throwOn404);
     }
 
-    private <T> T exec(Request request, Class<T> resultClass, int expectedCode) {
+    private <T> T exec(Request request, Class<T> resultClass, int expectedCode, boolean throwOn404) {
         try (Response response = client.newCall(request).execute()) {
             if (response.code() != expectedCode) {
+                if (response.code() == 404 && !throwOn404) {
```

Comment:
@vsevel Why have you differentiated here ? Can we already have the users who have the code expecting the exceptions if nothing is available on a given path ?

Reply:
@sberyozkin historically, any code other than the expected code would yield an exception. that was the default behavior. as I was writing the tests for the crud use case, it felt that getting a 404 when getting a secret on a non existent path should return a null instead of an exception. so I kept the old behavior for all calls, except for readSecret(), where a null seemed more natural.
how would you do it? what is your suggestion?

Reply:
Hi @vsevel if `404` is returned when no path actually exists in the secret engine then it is an exception IMHO (yes, it would also change the expectations for the existing users without a strong reaspn); if it returns 404 when nothing is available at that existing path then `null` is fine; I propose to keep the old behavior for the moment (and perhaps revisit with a dedicated issue ?)

Thanks

Reply:
hello @sberyozkin not sure I completely followed you.
when you read secrets from vault, you get a 404 if the path does not exist, and the secret (which is a Map<String,String>) if it does. if not keys are added to an existing path, I did not test but I think it returns an empty map.

so today the behavior in quarkus is:
 - path does not exist => exception
 - path does exist => map (which might be empty)

in the PR, for the readSecret operation, I changed the behavior to be:
 - path does not exist => null
 - path does exist => map (which might be empty)

do you want me to revert to the old behavior?

Reply:
@vsevel Thanks; yes the current behavior looks correct to me, as with this PR's change the difference between the non-existent path and no secrets being available at the existing path becomes `null` vs `empty Map`. I feel an exception is warranted if the users has used the path which has not been configured. Please revert this one, As I said, we can have a comprehensive discussion about the way the Vault responses are treated (for all the supported engines) in a dedicated issue and change as needed :-)

Reply:
ok. I will revert to the old behavior and squash all commits. then you can approve if no other comments.
I will do this tonight.

>  I feel an exception is warranted if a user has used the path which has not been correctly configured

minor correction: we are not talking about using a path that has not been configured correctly, but rather getting a non existent path that is passed as an argument.

if vault was a java object, it would look like this:
```
Map<String, Properties> vault = new HashMap<>()

vault.put("path1", new Properties());

Properties props2 = new Properties();
props2.put("mykey", "myvalue");
vault.put("path2", props2);

vault.get("path1") => empty properties object
vault.get("path2") => properties object with mykey=myvalue
vault.get("foobar") => null
```

so getting a null looks right from a java api perspective.

Reply:
Hi Vincent @vsevel that definitely looks good when we have a local map :-). But not IMHO when the `vault` secret engine interface represent a remote Vault server. That `foobar` in your example, it can actually be set via a custom Quarkus property, but the user forgot that the path was in fact `barfoo` :-).
So yeah, please revert it and I'll be happy to merge, thanks

Reply:
Hi Vincent @vsevel I was think while I was offline if I was unreasonable in asking you to revert this change, but eventually came to the conclusion the way it currently works is good. 
Returning to your java code example, I should say that the API designers often use the empty List or Map for a pure convenience purposes, to let the users avoid checking for `null` :-), as opposed to overloading the response value to mark an exceptional situation, such that the users can write (they might already be writing it this way now):
```
try {
    return vault.get("path1").get("secret1Key");
} catch (VaultException ex) {}
```
The new code would leak NPE and result in 500.
So IMHO we will be fine with keeping your original code idea in place :-) thanks

Reply:
404 changes have been reverted. I have squashed the commits. waiting for CI to go through. cheers.